File: pml/expander.py
"""PML macro expander — pre-evaluation pass that expands macro calls."""

from __future__ import annotations

from typing import Any

from pml.environment import Environment
from pml.errors import MacroExpansionDepthError
from pml.types import Expr, Macro, Symbol


class Expander:
    """Macro expander that runs before evaluation.

    Recursively expands macro calls in the AST. Non-hygienic:
    macro authors must use (gensym) for unique variable names.

    The expansion algorithm:
    1. If expr is not a list or is empty → return as-is
    2. If head is a Symbol resolving to a Macro → expand and re-expand result
    3. Otherwise → recursively expand all sub-expressions
    """

    def __init__(self, env: Environment, max_depth: int = 256) -> None:
        self.env = env
        self.max_depth = max_depth

    def expand(self, expr: Expr, depth: int = 0) -> Expr:
        """Expand a single expression, recursively expanding macro calls."""
        if depth > self.max_depth:
            raise MacroExpansionDepthError(
                f"Macro expansion exceeded maximum depth of {self.max_depth}"
            )

        # Atoms and empty lists pass through unchanged
        if not isinstance(expr, list) or len(expr) == 0:
            return expr

        head = expr[0]

        # Check if head is a macro
        if isinstance(head, Symbol):
            val = self.env.try_lookup(head.name)
            if isinstance(val, Macro):
                # Expand: substitute unevaluated args into macro body
                expanded = val.expand(expr[1:])
                # Re-expand the result (macros can produce more macro calls)
                return self.expand(expanded, depth + 1)

        # Not a macro call: recursively expand sub-expressions
        # But preserve special forms — don't expand inside quote/quasiquote
        if isinstance(head, Symbol):
            if head.name == "quote":
                return expr  # Don't expand inside quotes
            if head.name in ("lambda", "defmacro"):
                # Expand body but not parameter list
                expanded_body = [self.expand(sub, depth) for sub in expr[2:]]
                return [expr[0], expr[1]] + expanded_body
            if head.name in ("define",):
                # For (define (name params) body...), expand body only
                if len(expr) >= 3 and isinstance(expr[1], list):
                    expanded_body = [self.expand(sub, depth) for sub in expr[2:]]
                    return [expr[0], expr[1]] + expanded_body

        # Default: recursively expand all sub-expressions
        return [self.expand(sub, depth) for sub in expr]

    def expand_all(self, ast: list[Expr]) -> list[Expr]:
        """Expand macros in a list of top-level expressions."""
        return [self.expand(expr) for expr in ast]

File: src/pml/frontend/expander.cpp
// ═══════════════════════════════════════════════════════════════════════════════
// PML Expander — Macro Expansion Pass Implementation
// ═══════════════════════════════════════════════════════════════════════════════
//
// Matches the Python PML expander (pml/expander.py) exactly.
// ═══════════════════════════════════════════════════════════════════════════════

#include "expander.h"
#include "environment.h"  // from pml/evaluator — Environment with try_lookup()

#include <string>
#include <utility>

namespace pml {

// ═══════════════════════════════════════════════════════════════════════════════
// Constructor
// ═══════════════════════════════════════════════════════════════════════════════

Expander::Expander(std::shared_ptr<Environment> env_, int max_depth_)
    : env(std::move(env_))
    , max_depth(max_depth_) {}

// ═══════════════════════════════════════════════════════════════════════════════
// expand — expand a single expression recursively
// ═══════════════════════════════════════════════════════════════════════════════

Result<Expr> Expander::expand(const Expr& expr, int depth) const {
    // 1. Depth limit
    if (depth > max_depth) {
        return std::unexpected(
            macro_expansion_depth_error(SourceLocation{}, max_depth));
    }

    // 2. Atoms and empty lists pass through unchanged
    const auto* list = get_list(expr);
    if (!list || list->empty()) {
        return expr;
    }

    const Expr& head = (*list)[0];

    // 3. Check if head is a macro
    if (is_symbol(head)) {
        const auto& sym = std::get<Symbol>(head);
        auto val = env->try_lookup(sym.name);
        if (val.has_value() && is_macro(*val)) {
            const auto& macro_ptr =
                *std::get_if<std::shared_ptr<Macro>>(&*val);

            // Get args: all elements after head (expr[1:])
            std::vector<Expr> args;
            if (list->size() > 1) {
                args.reserve(list->size() - 1);
                for (size_t i = 1; i < list->size(); ++i) {
                    args.push_back((*list)[i]);
                }
            }

            // Expand: substitute unevaluated args into macro body
            Expr expanded = macro_ptr->expand(args);

            // Re-expand the result (macros can produce more macro calls)
            return this->expand(expanded, depth + 1);
        }
    }

    // 4. Not a macro call: recursively expand sub-expressions
    //    But preserve special forms — don't expand inside quote/quasiquote
    if (is_symbol(head)) {
        const auto& sym = std::get<Symbol>(head);

        if (sym.name == "quote") {
            return expr;  // Don't expand inside quotes
        }

        if (sym.name == "lambda" || sym.name == "defmacro") {
            // Expand body (expr[2:]) but not parameter list (expr[1])
            std::vector<Expr> result;
            result.reserve(list->size());
            result.push_back((*list)[0]);  // head (lambda/defmacro)
            result.push_back((*list)[1]);  // parameter list (unexpanded)
            for (size_t i = 2; i < list->size(); ++i) {
                auto expanded = expand((*list)[i], depth);
                if (!expanded) {
                    return std::unexpected(expanded.error());
                }
                result.push_back(std::move(*expanded));
            }
            return make_list(std::move(result));
        }

        if (sym.name == "define") {
            // For (define (name params) body...), expand body only
            if (list->size() >= 3 && is_list((*list)[1])) {
                std::vector<Expr> result;
                result.reserve(list->size());
                result.push_back((*list)[0]);  // define
                result.push_back((*list)[1]);  // function form (unexpanded)
                for (size_t i = 2; i < list->size(); ++i) {
                    auto expanded = expand((*list)[i], depth);
                    if (!expanded) {
                        return std::unexpected(expanded.error());
                    }
                    result.push_back(std::move(*expanded));
                }
                return make_list(std::move(result));
            }
            // Variable form (define name value) falls through to default
        }
    }

    // 5. Default: recursively expand all sub-expressions
    std::vector<Expr> result;
    result.reserve(list->size());
    for (const auto& elem : *list) {
        auto expanded = expand(elem, depth);
        if (!expanded) {
            return std::unexpected(expanded.error());
        }
        result.push_back(std::move(*expanded));
    }
    return make_list(std::move(result));
}

// ═══════════════════════════════════════════════════════════════════════════════
// expand_all — expand macros in a list of top-level expressions
// ═══════════════════════════════════════════════════════════════════════════════

Result<std::vector<Expr>> Expander::expand_all(
    const std::vector<Expr>& ast) const {
    std::vector<Expr> result;
    result.reserve(ast.size());
    for (const auto& expr : ast) {
        auto expanded = expand(expr);
        if (!expanded) {
            return std::unexpected(expanded.error());
        }
        result.push_back(std::move(*expanded));
    }
    return result;
}

}  // namespace pml
